src/details/desugar.cpp

```diff
@@ -0,0 +1,190 @@
+#include "ast.hpp"
+namespace minilua::details::ast {
+auto ForStatement::desugar() const -> DoStatement {
```

Comment:
I think desugaring needs some documentation about what code it generates. Maybe just an example is fine.

The code is very hard to understand (even though it is simple). Maybe this can also be split up into sections with some comments that show whats getting generated next (splitting into different functions is probably not possible).